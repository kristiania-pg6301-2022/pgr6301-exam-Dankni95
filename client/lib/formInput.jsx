import React from "react";

export function FormInput({label, value, onChangeValue}) {
    return (
        <div className="form-input">
            <label>
                <strong>{label}</strong>{" "}
                <input value={value} onChange={(e) => onChangeValue(e.target.value)}/>
            </label>
        </div>
    );
}

export function FormTextArea({label, value, onChangeValue}) {
    return (
        <div className="form-input">
            <label>
                <strong>{label}</strong>{" "}
                <textarea value={value} onChange={(e) => onChangeValue(e.target.value)}/>
            </label>
        </div>
    );
}

export function FormTextOptions({label, value, onChangeValue}) {
    return (
        <div className="form-input">
            <div>
                <select value={value} onChange={(e) => onChangeValue(e.target.value)}>
                    <option value="Politics">Politics</option>
                    <option value="Economics">Economics</option>
                    <option value="Sport">Sport</option>
                    <option value="Health">Health</option>
                </select>
            </div>
            <strong>{label}</strong>{" "}
        </div>
    );
}
