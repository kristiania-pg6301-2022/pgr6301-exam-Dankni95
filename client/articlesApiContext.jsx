import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";

export const ArticlesApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async listArticles(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async createArticle(article) {
    return await postJSON("/api/articles", article);
  },
  async deleteArticle(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});
