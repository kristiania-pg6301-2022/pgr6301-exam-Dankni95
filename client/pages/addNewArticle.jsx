import React, {useContext, useState} from "react";
import {useNavigate} from "react-router-dom";
import {FormInput, FormTextArea, FormTextOptions} from "../lib/formInput";
import {ArticlesApiContext} from "../articlesApiContext.jsx";

export function AddNewArticle({user}) {
    const {createArticle} = useContext(ArticlesApiContext);
    const [title, setTitle] = useState("");
    const [category, setCategory] = useState("Politics");
    const [content, setContent] = useState("");

    const navigate = useNavigate();

    function handleSubmit(event) {
        event.preventDefault();
        const {name} = user.openid
        createArticle({title, date: new Date().toLocaleDateString(), category, content, name});
        navigate("/");
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>Add new article</h1>

            <FormInput label={"Title:"} value={title} onChangeValue={setTitle}/>
            <FormTextOptions label={"Category: "} value={category} onChangeValue={setCategory}/>
            <FormTextArea label={"Content:"} value={content} onChangeValue={setContent}/>
            <div>
                <button>Submit</button>
            </div>
        </form>
    );
}
