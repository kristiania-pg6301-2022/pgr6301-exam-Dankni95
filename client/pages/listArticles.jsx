import React, {useContext, useState, useEffect} from "react";
import {ArticlesApiContext} from "../articlesApiContext.jsx";
import {useNavigate} from "react-router-dom";
import {useLoading} from "../useLoading";

export async function handleClick(article, deleteArticle) {
    await deleteArticle(article)
    useNavigate()
}


function Article({article: {title, date, category, content, name}, user}) {
    const {deleteArticle} = useContext(ArticlesApiContext)

    return (
        <div>
            <h6>{category}</h6>
            <h1>{"title" + title}</h1>
            <h6>{date}</h6>
            {user?.google || user?.openid ? <h4>{content}</h4> : "Sign in to read the article"}
            <h6>By: {name}</h6>
            {user.openid && user?.openid?.name === name ?
                <button onClick={(event) => handleClick({del: title}, deleteArticle)}>Edit</button> : ""}
        </div>
    );
}


export function ListArticles({user}) {
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");


    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(categoryQuery);
    }

    const [articles, setArticles] = useState("");
    useEffect(() => {
        const ws = new WebSocket(window.location.origin.replace(/^http/, "ws") + "/");
        ws.onmessage = (event) => {
            const {name, title, content, date, category} = JSON.parse(event.data);
            setArticles((articles) => [...articles, {name, title, content, date, category}]);
        };
    }, []);




    return (
        <div>
            <h1>Articles</h1>

            <div>
                <form onSubmit={handleSubmitQuery}>
                    <label>
                        Category:
                        <select name={"Category"} value={categoryQuery}
                                onChange={(e) => setCategoryQuery(e.target.value)}
                        >
                            <option value="">All</option>
                            <option value="Politics">Politics</option>
                            <option value="Economics">Economics</option>
                            <option value="Sport">Sport</option>
                            <option value="Health">Health</option>
                        </select>
                        <button>Select</button>
                    </label>
                </form>
            </div>
            {Object.keys(articles).map((keyName, i) => (
                <Article key={articles[keyName].title} article={articles[keyName]} user={user}/>
            ))}
        </div>
    );
}
