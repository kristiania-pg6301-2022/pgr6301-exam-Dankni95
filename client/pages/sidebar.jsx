import React, {useContext, useState, useEffect} from "react";
import {ArticlesApiContext} from "../articlesApiContext.jsx";
import "./sidebar.css"

export default (user) => {
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");
    const [ws, setWs] = useState();


    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(e.target.value);
    }
    const [articles, setArticles] = useState("");

    useEffect(() => {
        const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
        ws.onopen = () => {
            ws.send(JSON.stringify({category: categoryQuery}))
        }

        setWs(ws)
    }, [category]);


    return (
        <div>
            <h2>Articles</h2>

            <div>
                <div>
                    <form>
                        <h1> Category:</h1>
                        <button onClick={(e)=>{
                            handleSubmitQuery(e)
                        }} value="">All</button>
                        <button value="Politics" onClick={(e)=>{
                            handleSubmitQuery(e)
                        }} >Politics</button>
                        <button value="Economics" onClick={(e)=>{
                            handleSubmitQuery(e)
                        }} >Economics</button>
                        <button value="Sport" onClick={(e)=>{
                            handleSubmitQuery(e)
                        }} >Sport</button>
                        <button value="Health" onClick={(e)=>{
                            handleSubmitQuery(e)
                        }} >Health</button>
                    </form>
                </div>
            </div>
        </div>
    );

};