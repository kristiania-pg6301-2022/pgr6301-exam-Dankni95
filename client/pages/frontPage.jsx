import {Link, Navigate} from "react-router-dom"
import React from "react"
import {ListArticles} from "./listArticles";
import Sidebar from "./sidebar";
import {Card} from "@mui/material";


export function FrontPage({user}) {

    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));


    return (
        <>
            <div className="left-side"><Sidebar  socket={ws}/></div>
            <main  style={{ padding: "1rem 1rem 1rem 1rem", width:"98%" }} className="main"><ListArticles user={user} ws={ws} /></main>
            <footer className="section coral">Â© 2022 Kristiania Avis</footer>
        </>
    )
}
