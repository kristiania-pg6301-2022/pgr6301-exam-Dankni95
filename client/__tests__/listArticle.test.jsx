import { ListArticles } from "../pages/listArticles";

import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { ArticlesApiContext } from "../articlesApiContext.jsx";
import {logPlugin} from "@babel/preset-env/lib/debug";

describe("ListMovies component", () => {
  it("shows loading screen", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<ListArticles />, domElement);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("shows movies", async () => {
    const articles = [{ title: "ar 1" }, { title: "ar 2" }];
    const domElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <ArticlesApiContext.Provider value={{ listMovies: () => articles }}>
          <ListArticles />
        </ArticlesApiContext.Provider>,
        domElement
      );
    });

    expect(
      Array.from(domElement.querySelectorAll("h6")).map((e) => e.innerHTML)
    ).toEqual(["movie 1", "movie 2"]);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("queries by country", async () => {
    const domElement = document.createElement("div");
    const listMovies = jest.fn(() => []);
    await act(async () => {
      ReactDOM.render(
        <ArticlesApiContext.Provider value={{ listMovies }}>
          <ListArticles />
        </ArticlesApiContext.Provider>,
        domElement
      );
    });
    Simulate.change(domElement.querySelector("#country-query"), {
      target: { value: "Ukraine" },
    });
    await act(async () => {
      await Simulate.submit(domElement.querySelector("form"));
    });
    expect(listMovies).toHaveBeenCalledWith({
      country: "Ukraine",
    });
  });

  it("shows error message", async () => {
    const domElement = document.createElement("div");
    await act(async () => {
      const listMovies = () => {
        throw new Error("Something went wrong");
      };
      ReactDOM.render(
        <ArticlesApiContext.Provider value={{ listMovies }}>
          <ListArticles />
        </ArticlesApiContext.Provider>,
        domElement
      );
    });

    expect(domElement.querySelector("#error-text").innerHTML).toEqual(
      "Error: Something went wrong"
    );
    expect(domElement.innerHTML).toMatchSnapshot();
  });
});
