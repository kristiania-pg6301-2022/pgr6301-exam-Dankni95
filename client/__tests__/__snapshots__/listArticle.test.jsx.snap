// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies component shows loading screen 1`] = `"<div><h1>Articles</h1><div><form><label>Category:<select name=\\"Category\\"><option value=\\"\\">All</option><option value=\\"Politics\\">Politics</option><option value=\\"Economics\\">Economics</option><option value=\\"Sport\\">Sport</option><option value=\\"Health\\">Health</option></select><button>Select</button></label></form></div></div>"`;
